### PR TITLE
Find GMSH executable.

### PR DESCRIPTION
**Pull request by luca-heltai**
I'm not sure this is the correct way to proceed for an executable to be found. Maybe @tamiko  has a different opinion... 

:)

This is a prequel to another PR, that creates a deal.II triangulation from a `TopoDS_Edge` object of OpenCASCADE, using gmsh internally.

### CODE REVIEW

**Comment by luca-heltai** [comment]
Linking against gmsh requires you to install the development version, compile it, etc. While this does not, and don't need anything else apart from the possibility to execute gmsh on temporary files.

**Comment by davydden** [comment]
> Linking against gmsh requires you to install the development version, compile it, etc.

that's not too difficult, especially if you don't do it by hand. But I understand your concerns.

I think we need more opinions on using Gmsh via executable as opposed to the library. @dealii/dealii ping.

**Comment by tjhei** [comment]
gmsh has an API so using it as a library would be neat (instead of reading/writing files to disk). I can see that this is a lot more work, though. We already have ways to read in .msh files, but one would need to write wrappers to convert the in-memory representation to a ``Triangulation`` to use the API.

**Comment by luca-heltai** [comment]
We have code that uses the gmsh executable to generate meshes from an OpenCASCADE object. I wanted to include this in the library, without having to write (yet another) library interface.

We can always add the interface to the API at a later time. Why are you so concerned about using an executable?

**Comment by bangerth** [comment]
I'd be ok with calling the executable. There is nothing that stops us from using the API at a later time, but that's clearly a superset of functionality from what's necessary for simple things and it's (presumably) substantially more complicated for users to install the API instead of just the executable.

File: cmake/modules/FindGMSH.cmake [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+# This module exports
+#
+#   GMSH_EXECUTABLE
+#
+
+SET(GMSH_DIR "" CACHE PATH "An optional hint to a GMSH installation containing the gmsh executable")
+SET_IF_EMPTY(GMSH_DIR "$ENV{GMSH_DIR}")
+
+DEAL_II_FIND_FILE(GMSH_EXE gmsh${CMAKE_EXECUTABLE_SUFFIX}
+  HINTS ${GMSH_DIR}
```
Comment (bangerth) [inline cmake/modules/FindGMSH.cmake]: Are `/usr` and `/usr/local` automatically `HINT`s in this context?

File: include/deal.II/base/config.h.in [hunk 2]
```diff
@@ -192,6 +193,14 @@
     >=  \
     (major)*10000 + (minor)*100 + (subminor))
 
+
+/*
+ * gmsh:
+ */
+#ifdef DEAL_II_WITH_GMSH
+#  define GMSH_EXECUTABLE @GMSH_EXECUTABLE@
```
Comment (bangerth) [inline include/deal.II/base/config.h.in]: Maybe call it `DEAL_II_GMSH_EXECUTABLE` to not tread on other packages' namespaces?

File: cmake/modules/FindGMSH.cmake (hunk 1 above)
Comment (luca-heltai) [reply cmake/modules/FindGMSH.cmake]: yes

File: include/deal.II/base/config.h.in (hunk 2 above)
Comment (luca-heltai) [reply include/deal.II/base/config.h.in]: This is consistent with all other packages. We don't set `DEAL_II_PETSC_INCLUDE_DIRECTORIES`, but `PETSC_INCLUDE_DIRECTORIES`. This is a requirement for any `FindPACKAGE.cmake` in `cmake` standards, as far as I know, and this is how we do it for all other packages in `cmake/modules/FindXXX.cmake`.
This is done through `cmake/macros/macro_deal_ii_package_handle.cmake`.

**Comment by luca-heltai** [comment]
Just as a hint of where we are going:


droplet

File: include/deal.II/base/config.h.in (hunk 2 above)
Comment (luca-heltai) [reply include/deal.II/base/config.h.in]: Oh, I see, maybe you mean the define inside the code, not inside cmake variables... :P I'll do that.

Comment (bangerth) [reply include/deal.II/base/config.h.in]: Yes, correct. The name of the `@...@` expanded variable doesn't matter -- it never escapes cmake. But the `#define`d variable can conflict with other packages.

**Comment by luca-heltai** [comment]
ping @tamiko

**Comment by luca-heltai** [comment]
The failing build does not seem to be related to this PR.

**Comment by masterleinad** [comment]
> The failing build does not seem to be related to this PR.

I agree. Would you just rebase and push again to restart the tester?

**Comment by tamiko** [comment]
@luca-heltai I am fine with the executable integration. (Though I am not a fan of calling executable from within the library, meaning you will lose karma points :-P)

**Comment by luca-heltai** [comment]
Sure, API is better. But I'll leave that to someone else, this time. I have many other things in my todo list, and studying yet another API is not there...

**Comment by luca-heltai** [comment]
Merge?

**Pull request merged**